src/modules/roc_core/target_linux/roc_core/timer.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <errno.h>
+#include <sys/timerfd.h>
+#include <unistd.h>
+
+#include "roc_core/errno_to_str.h"
+#include "roc_core/panic.h"
+#include "roc_core/timer.h"
+
+namespace {
+
+struct itimerspec infinity_deadline() {
+    struct itimerspec new_value;
+    new_value.it_interval.tv_sec = 0;
+    new_value.it_interval.tv_nsec = 0;
+    new_value.it_value.tv_sec = 0;
+    new_value.it_value.tv_nsec = 0;
+
+    return new_value;
+}
+
+struct itimerspec immediately_deadline() {
+    struct itimerspec new_value;
+    new_value.it_interval.tv_sec = 0;
+    new_value.it_interval.tv_nsec = 0;
+    new_value.it_value.tv_sec = 0;
+    new_value.it_value.tv_nsec = roc::core::Nanosecond;
+
+    return new_value;
+}
+
+struct itimerspec finity_deadline(roc::core::nanoseconds_t deadline) {
+    struct itimerspec new_value;
+    new_value.it_interval.tv_sec = 0;
+    new_value.it_interval.tv_nsec = 0;
+    new_value.it_value.tv_sec = (time_t)(deadline / roc::core::Second);
+    new_value.it_value.tv_nsec = deadline % roc::core::Second;
+
+    return new_value;
+}
+
+struct itimerspec convert_deadline(roc::core::nanoseconds_t deadline) {
+    if (deadline < 0) {
+        return infinity_deadline();
+    } else if (deadline == 0) {
+        return immediately_deadline();
+    } else {
+        return finity_deadline(deadline);
+    }
+}
+
+} // namespace anonymous
+
+namespace roc {
+namespace core {
+
+Timer::Timer()
+    : timerfd_(timerfd_create(CLOCK_MONOTONIC, 0))
+    , deadline_(0)
+    , waiting_flag_(false)
+    , sem_(1) {
+    if (timerfd_ == -1) {
+        roc_panic("timer: timerfd_create(): %s", errno_to_str().c_str());
+    }
+}
+
+Timer::~Timer() {
+    int res = close(timerfd_);
+    if (res == -1) {
+        roc_panic("timer: close(): %s", errno_to_str().c_str());
+    }
+}
+
+bool Timer::try_set_deadline(nanoseconds_t new_deadline) {
+    if (!deadline_.try_store(new_deadline)) {
+        return false;
+    }
+
+    sem_.wait();
```

Comment:
Could you explain what are we trying to achieve using this sem_.wait() call?

Reply:
Is synchronization try_set_deadline and wait_deadline, because it have complexity. While one thread execute wait_deadline, other execute try_set_deadline - significant deadline can be drop. It was in tests before. Semaphore guard significant part of code, where concurrent threads can corrupt deadline.

Reply:
But do you really understand the requirement of making try_set_deadline() lock-free (except the syscall)? This issue requires some basic understanding of this stuff.

I see that you're using sem_ to create a critical section, but why did you end up with this solution? The current implementation of try_set_deadline() may be blocked by other try_set_deadline() or wait_deadline() calls, which is exactly what we want to avoid for the reasons I mentioned earlier.

Reply:
If you want lock-free, this functions must look like:

try_set_deadline
{
syscall_set(new_deadline);
}

wait_deadline
{
syscall_wait();
}

Other ideas as check on waiting and other is complexity and come as critical section. As I say, without it we loose significant deadlines and catch deadlock on wait_deadline.

We can't do so many actions as atomic operation.

Reply:
I feel that lock-free implementation is possible, since there is no big difference from cond-based Timer implementation, which is lock-free. However I don't have the full algorithm in mind so far. I'll look into it a bit later, after finishing more urgent issues of 0.2 release. Ping me if you'll have any ideas.